### ISSUE TITLE
Saved images get epoch milliseconds as name like 

### ISSUE THREAD

**Issue by Mnyromyr** [signalapp/Signal-Android#10142]
- [x ] I have searched open and closed issues for duplicates
- [x ] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x ] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Images received in a conversation get the epoch milliseconds as their name when stored in local storage.

### Steps to reproduce
- get image in conversation
- tap image to show it larger
- tap save from the dot menu
- ack the local storage warning
- Signal says "saved to media"

**Actual result:** file is actually stored in local storage ~/Pictures/ and named something like 1603482657869.jpg which looks like a Unix epoch millisecond timestamp

**Expected result:** file name should be more user friendly, eg. "signal-2020-10-26-201624.jpeg" like before

** Additional info:
- probably related to issue 10140 which appeared around the same time
- seems to have appeared roughly somewhere between 2020-10-19 and 2020-10-23

### Device info

**Device:** Samsung S7 / Huawei Little 9
**Android version:** 8.0.0 / unknown
**Signal version:** 4.74.3 / 4.74.3

**Comment by Genis-Sage** [comment]
hey @Mnyromyr 
I could not reproduce the behaviour on my phone
and also, 1603482657869 converted would be the year 52782

could you save a picture again and then create a debug log

also, does this happen with all you images, or only with certain ones?
if it happens only for certain ones, what do they have in common, which the others don't have? (like a special file type, etc)

**Comment by Mnyromyr** [comment]
0. The Signal  version on my Samsung is now 4.75.2, but the issue is still there.
1. I said milliepoch on purpose: 1603482657869 are milliseconds, hence it's 2020-10-23T19:50:57.869Z.
 (The Samsung file manager displays a wrong file date with year 52796, though.)
2. I tried a couple of pictures/photos from various conversations with various people and groups, it's always the same broken behaviour now (even sending a png results in wrongly named jpg). PDFs are not affected. 
3. The problem does _not_ occur with the Linux desktop app v1.34.5, although the naming scheme is quite incoherent: Images seem to retain an original name here (I suppose) while photos get named signal-2020-10-23-184256_002.jpeg etc. 
4. Debug log: https://debuglogs.org/59ecbd2d11b7309f53f75f81695d5b3fa26c04e264f7313a1a027521a5aee3b6

**Comment by greyson-signal** [comment]
Really strange. If we auto-generate a filename, it'll always be like `signal-2020-10-26-201624.jpeg`. And unfortunately I can't reproduce this on any of my devices.

Is it possible that the photos being sent to you have a filename of a timestamp? Maybe try taking and sending a photo to yourself and saving that. You can compare the filename of the captured photo to the filename of the photo saved from signal.

**Comment by Mnyromyr** [comment]
If I send myself a local image to "Not to self", its name is lost when saving it from this channel in Signal.

I pestered some family members and got this:
(0) Samsung S7 / Android 8 / Signal 4.75.4 => broken filename
(1) Huawei Little 9 / Android 7 /  Signal 4.74.3 => broken filename
(2) Samsung Galaxy Note 8 / Android 9 / Signal 4.74.3 => normal filename

So my guess would be a problem related to the Android version?!

I attached a screenshot of the SAmsung file manager showing details of such a broken saved file.

**Comment by Mnyromyr** [comment]
![signal-2020-10-29-173647](https://user-images.githubusercontent.com/22755723/97619518-9a9a2100-1a20-11eb-8463-f6a66d6715b6.jpeg)

**Comment by Genis-Sage** [comment]
@Mnyromyr 
sorry for the misunderstanding, I somehow jump over the fact, that you wrote milliseconds,
and my epectation was seconds

@greyson-signal 
i could reproduce the behaviour now

I send a picture from my Android 10 main phone to my Android 6 spare phone
and then saved the picture on both devices
on Android 10, the file has the correct name and modified date
on Android 6, I get the milliseconds name, but the modified date is correct (year 2020 and not year 52798)

so it seems, something changed from Android 8 (and older) to Android 9 and newer

here's a debug log of my Android 6 device, like half a minute after saving the picture:
https://debuglogs.org/f1114ceec156d0b6387bc6b57f578a5d72074a5503f3cce9630c43b1a6c01fff

**Comment by helge79** [comment]
I also recently experienced this issue with 4.75.4 on a Xiamoi Mi A1 with Lineage OS 16 (Android 9, no Google Services installed). I am using the APK from signal.org. This definetely worked before

One additional observation: the file name is the epoch time of the creation date of the file and not the one when the picture was taken. I also noticed videos are stored as *.3gp (and not *.mp4 anymore).

**Comment by unhammer** [comment]
For anyone who already exported and deleted, you can [exiftool](https://exiftool.org/) to get the original date out:

```sh
echo '${CreateDate}_${FileName}' > fmtfile.txt
for f in *3gp; do mv "$f" "$(exiftool -d "%Y-%m-%d_%H.%M.%S" -p fmtfile.txt "$f")"; done
```

**Comment by fumiakiy** [comment]
This issue should be resolved by merging [this pull request](https://github.com/signalapp/Signal-Android/pull/10283). In particular by [setting the meta data here](https://github.com/signalapp/Signal-Android/pull/10283/files#diff-226ddd30ae587f078b07d87eaf520b8d96baa4d323d7ac7b4d1df891f7adabdfR226).

The root cause is that some flavors of some versions of Android has a content resolver implementation that generates the file name by `Date.now()` (or an equivalent of it), presumably to avoid a file name conflicts with an existing one.

[The meta data that gives the content resolver a hint about the file name](https://github.com/signalapp/Signal-Android/pull/10283/files#diff-226ddd30ae587f078b07d87eaf520b8d96baa4d323d7ac7b4d1df891f7adabdfR226) however, will cause another issue. Most content resolvers that respect the meta data does nothing to prevent from the given file name from conflicts. They simply returns null to the call to `insert` method that ultimately leads crashes reported in #10159. [The pull request](https://github.com/signalapp/Signal-Android/pull/10283) fixes the issue too by checking and regenerating new files until the file name does not conflict any more.

**Comment by joha1** [comment]
I've got the same issue (Correct signal-2021-01-06-xxxxx.jpeg filename, but shows up in gallery in the year 52987).
In my case it's IPhone SE/iOS 12.4.1/Signal 5.0.3.0 (and a bunch of earlier version) -> Samsung A7/Android 10/Signal 5.0.8 which results in borked time stamps, whereas the reverse (Android to iOS) with the same two devices works.
Also, as in  Mnyromyrs case, the phones in question also have their language set to German, so could there be some DMY vs. YMD date convention issues involved?

**Comment by waandersjethro** [comment]
I've the same problem. Using Signal version 5.1.9, Android 7 and a Motorola Moto G4 Plus.
Dates like "1 September 53011"  are shown.

**Comment by skupfer** [comment]
Same issue here. 

i.e. `signal-2020-12-26-174135.jpg` but the date is "8. Februar 53014 06:34". Several chats affected.

Signal 5.0.8 (5.1.9 after updating)
Android 10
Galaxy A6

After updating to Signal 5.1.9 and a freshly saved picture "13. Februar 53014 20:58" and filename `signal-2020-12-12-224539.jpg"

**Comment by freddyberg96** [comment]
I have the same issue with both the filename being wrong, something like 1610308141567 and the date being wrong with something like year 52998.

Signal 5.0.8 and still after Signal 5.1.9
Oneplus 3
LineageOS 15.1 with microG (english but german locals)
Android 8.1.0

**Comment by mike-tuxedo** [comment]
I also can confirm a wrong date of media, when it is saved outside of Signal. 
For exampl an image, in Signal the date was 15.01.2021, saved outside it was 20.02.53019
Device: Samsung Galaxy S7
Android: 8.0.0
Signal: 5.1.9

**Comment by lemonnuggets** [comment]
Same problem here. I'm not sure about how all this works but it appears as though Signal is saving the epoch time in milliseconds while the rest of the device expects the time in seconds. 
This is a really annoying issue as images I saved a few days ago will still show up on the top in any window where I might be trying to select any image on my device. For some inexplicable reason, after a few days of having this weird date metadata, it changes back to the expected date.
When I say metadata I don't really know exactly where this metadata is being stored. On my phone when I check (on Google Photos) the date that this photo was saved, the date it shows corresponds to the current epoch time in milliseconds. But when I sent the same image to my laptop to check the file metadata, there was no date specified. So I don't really know where Google Photos is getting the date from but since I haven't faced this problem with any other application I think its Signal that is misbehaving.
Also definitely feels like this should be its own separate issue.
Device: Moto G5S Plus
Android: 8.1.0
Signal: 5.2.3

**Comment by cjee21** [comment]
Experiencing the same issue regarding saved image date as well as filename. I have tested on Android 9 (Sony Xperia) and Android 10 (Android One). The issue only occurs on the Android 9 phone.

<img src="https://user-images.githubusercontent.com/77721854/105150151-5ca02b00-5b3f-11eb-8452-e67b55506473.png" height="500"> <img src="https://user-images.githubusercontent.com/77721854/105150181-645fcf80-5b3f-11eb-9bd2-905871772773.png" height="500">
_Left: Android 9, Right: Android 10_

I suspect the date issue is caused by the use of milliseconds instead of seconds when setting the metadata for Android's MediaStore. The reason why some people have the date fix itself after a few days could be because Android's MediaStore has re-scanned the image and updated the date in MediaStore database to the actual date modified of the file.

Upon investigation, I found that [Signal's code](https://github.com/signalapp/Signal-Android/blob/09c5ba1c8294c7188ac5deaa30b08249b2f0a25f/app/src/main/java/org/thoughtcrime/securesms/util/SaveAttachmentTask.java#L184) uses milliseconds for DATE_ADDED and DATE_MODIFIED. I believe it should be using seconds as described in [Android's source code](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android10-mainline-media-release/core/java/android/provider/MediaStore.java#959). Only DATE_TAKEN is using milliseconds on Android's MediaStore. The issue can potentially be fixed by changing the two lines of code to the following:
`contentValues.put(MediaStore.MediaColumns.DATE_ADDED, System.currentTimeMillis() / 1000);
 contentValues.put(MediaStore.MediaColumns.DATE_MODIFIED, System.currentTimeMillis() / 1000);`

I agree that this date issue should be another issue that's separate from the filename issue. I am also experiencing the filename issue where the filename is epoch in milliseconds instead of a proper filename. I also notice that PNG files get wrongly saved with JPG extension on Android 9 when this filename issue happens.

**Comment by overlinden** [comment]
Same issue on my device.
I use Android 9 and Signal 5.2.3.
Please merge the corresponding pull request.

**Comment by cjee21** [comment]
These two bugs are still present in Signal 5.3.7 as expected since no code changes were made.

I hope the bugs can be fixed soon because not only does saved media from signal remain at the top of the phone gallery forever (due to the wrong date for MediaStore bug) but also GIF and PNG images all get saved wrongly with JPG file extension while MP4 videos get saved wrongly with 3GP file extension (due to the filename bug) thus requiring them to be manually renamed.

These two bugs should be separate and unrelated. The pull request for fixing the filename bug does not seem to be able to fix the wrong date in gallery bug.

Operating System: Android 9
Signal version: 5.3.7 from Play Store

**Comment by MaDill** [comment]
Both issues show up also on my phone.

BQ Aquaria X
Stock ROM Android 8.1.0 (2394, 2.11.0_20191121-1509)
Signal 5.3.7.1 (780, 78007) from Google Play Store
Locale  de_DE

**Comment by MJNL** [comment]
Hello, I've created a Github account just to reply to this issue with a possible workaround for the incorrect date/year in your Android gallery. Note that I have no programming knowledge but maybe it's some help for users or a hint for developers to work out further.

Like some of you I was also annoyed by having the downloaded images from Signal always on top with incorrect dates and years like 53071, So I tried to hide the complete Signal Pictures folder with a .nomedia file. Then I changed my mind, because I still wanted to see those images in my gallery, and suddenly the issue was solved (for previous downloaded images only, new images will have the incorrect date again).

Note that this does NOT change the file name itself, but it gets your images in the right order again, with the correct date and timestamp from when you downloaded the picture from Signal to your internal storage. I think these are two separate issues in one thread.

Anyway, what worked for me:

1 Connect your phone to your pc,
2 Create a blank .nomedia file and put this in the internal storage/Pictures folder
3. Open the gallery on your phone. You will get an alert that there is a .nomedia file detected, ignore and close the album
4. Delete the .nomedia from the folder again on your pc
5. Open your gallery on your phone again and smile. 

Before
![before](https://user-images.githubusercontent.com/78639031/107115029-0f73c700-686a-11eb-873a-9eacd06ade28.png)

After
![after](https://user-images.githubusercontent.com/78639031/107115028-0edb3080-686a-11eb-8c11-f1a2ff01e62d.png)


NOTE: The issue is completely solved after I updated my Android to version Android 11, so it's probably a bug for lower Android versions only. 


Hope this is any help!

**Comment by cjee21** [comment]
I have done some tests regarding Android MediaStore and can now confirm that changing two lines of code as mentioned in my [previous comment](https://github.com/signalapp/Signal-Android/issues/10142#issuecomment-763452512) does indeed fix the wrong date in photo gallery issue. The reason is because Android's MediaColumns for DATE_ADDED and DATE_MODIFIED is supposed to be in seconds and NOT milliseconds.

In short, just a small and simple change is required [here](https://github.com/signalapp/Signal-Android/blob/e6e8786d86aa3b0b340c1df14d25e6fec129425b/app/src/main/java/org/thoughtcrime/securesms/util/SaveAttachmentTask.java#L165) to fix the date bug as shown below:
```diff
-     contentValues.put(MediaStore.MediaColumns.DATE_ADDED, System.currentTimeMillis());
-     contentValues.put(MediaStore.MediaColumns.DATE_MODIFIED, System.currentTimeMillis());
+     contentValues.put(MediaStore.MediaColumns.DATE_ADDED, System.currentTimeMillis() / 1000);
+     contentValues.put(MediaStore.MediaColumns.DATE_MODIFIED, System.currentTimeMillis() / 1000);
```

On the other hand, the filename and file extension issue should be resolved with [this pull request](https://github.com/signalapp/Signal-Android/pull/10283/files#diff-226ddd30ae587f078b07d87eaf520b8d96baa4d323d7ac7b4d1df891f7adabdfR226) by setting Android's MediaColumn for DATA which is required on Android versions lower than 10 for proper filenames.

**Comment by Sgn-32** [comment]
Commit to 5.4 beta: https://github.com/signalapp/Signal-Android/commit/b93599954883e8129c1fead23a8aafaaa7dd650b

**Comment by cjee21** [comment]
I can confirm that on Android 9 with Signal version 5.4.1, images and videos are now saved with the correct time, date, year, filename and file extension.

Tested on:
Signal version: 5.4.1 from Play Store
OS: Stock Android 9
Phone: Sony Xperia
Gallery app: Google Photos and Sony Album

**Comment by cody-signal** [comment]
Going to close this, if you're on 5.4.1 or later and continue to experience the issue, let us know.

**Comment by Mnyromyr** [comment]
As the original submitter I can confirm that this is fixed with Signal 5.4.7.

**Comment by waandersjethro** [comment]
Tested with Signal 5.4.12, Moto G. Works fine now! Thanks.
